File: app/lib/features/club/models/club_month.dart
/// Клуб месяца. Соответствует server/src/models/ClubMonth.js.
///
/// Один документ на месяц. Каждое 1-е число стартует новый клуб,
/// предыдущий уходит в архив на 21 день (archiveUntilDate).
class ClubMonth {
  const ClubMonth({
    required this.id,
    required this.month,
    required this.year,
    required this.bookId,
    required this.title,
    required this.author,
    required this.startsAt,
    required this.endsAt,
    required this.archiveUntilDate,
    required this.isActive,
    required this.participantCount,
    required this.messageCount,
    this.partSchedule = const [],
    this.pinnedMessageId,
  });

  /// MongoDB ObjectId как строка.
  final String id;

  /// Месяц (1..12) и год.
  final int month;
  final int year;

  /// Ссылка на Book — детали через `Book` модель в response /api/club/current.
  final String bookId;

  /// Денормализованные поля из Book (для скорости — сервер кладёт их сюда при создании).
  final String title;
  final String author;

  /// Период действия клуба.
  final DateTime startsAt;
  final DateTime endsAt;

  /// Дата окончания архивного доступа = endsAt + 21 день.
  /// Используется чтобы решить: показывать ли клуб юзеру с истёкшей подпиской.
  final DateTime archiveUntilDate;

  /// Cron-флаг активного клуба. true когда startsAt <= now < endsAt.
  final bool isActive;

  /// Статистика (обновляется сервером триггерами).
  final int participantCount;
  final int messageCount;

  /// Расписание открытия частей. Пустой массив — все части доступны со старта.
  final List<PartSchedule> partSchedule;

  /// ID закреплённого сообщения (1 на клуб, только Анна закрепляет).
  final String? pinnedMessageId;

  factory ClubMonth.fromJson(Map<String, dynamic> json) {
    final scheduleRaw = json['partSchedule'];
    final schedule = scheduleRaw is List
        ? scheduleRaw
            .map((e) => PartSchedule.fromJson(e as Map<String, dynamic>))
            .toList(growable: false)
        : const <PartSchedule>[];

    return ClubMonth(
      id: (json['_id'] ?? '').toString(),
      month: (json['month'] as num?)?.toInt() ?? 1,
      year: (json['year'] as num?)?.toInt() ?? 2026,
      bookId: (json['bookId'] ?? '').toString(),
      title: (json['title'] ?? '').toString(),
      author: (json['author'] ?? '').toString(),
      startsAt: _parseDate(json['startsAt']) ?? DateTime.now(),
      endsAt: _parseDate(json['endsAt']) ?? DateTime.now(),
      archiveUntilDate: _parseDate(json['archiveUntilDate']) ?? DateTime.now(),
      isActive: json['isActive'] == true,
      participantCount: (json['participantCount'] as num?)?.toInt() ?? 0,
      messageCount: (json['messageCount'] as num?)?.toInt() ?? 0,
      partSchedule: schedule,
      pinnedMessageId: json['pinnedMessageId']?.toString(),
    );
  }

  /// Есть ли закреплённое сообщение в этом клубе.
  bool get hasPinnedMessage =>
      pinnedMessageId != null && pinnedMessageId!.isNotEmpty;

  /// Клуб полностью закрыт (даже архив).
  bool isArchiveClosed(DateTime now) => now.isAfter(archiveUntilDate);

  /// Клуб ещё не начался (будущий).
  bool isFuture(DateTime now) => now.isBefore(startsAt);
}

/// Расписание открытия одной части (используется в табе «Разборы»).
class PartSchedule {
  const PartSchedule({required this.partNumber, required this.opensAt});

  final int partNumber;
  final DateTime opensAt;

  factory PartSchedule.fromJson(Map<String, dynamic> json) {
    return PartSchedule(
      partNumber: (json['partNumber'] as num?)?.toInt() ?? 1,
      opensAt: _parseDate(json['opensAt']) ?? DateTime.now(),
    );
  }
}

DateTime? _parseDate(dynamic raw) {
  if (raw is String && raw.isNotEmpty) return DateTime.tryParse(raw);
  return null;
}

File: server/src/models/ClubMonth.js
const mongoose = require('mongoose');

/**
 * Клуб месяца — главная сущность Фазы 4.
 *
 * Один месяц = одна книга. Каждое 1-е число стартует новый клуб,
 * предыдущий уходит в архив. Доступ к архивному клубу — 21 день
 * после endsAt для пользователей чья подписка истекла (см. MASTER 4.37).
 *
 * Связь с Book:
 * - bookId — ссылка на конкретный аудиоразбор (4 части)
 * - Book.isPartOfClub=true и Book.clubMonth='YYYY-MM' дают обратную связь
 *
 * Расписание открытия частей:
 * - partSchedule — массив дат открытия для каждой части (1..N)
 * - При запросе клуба фронтенд сравнивает с текущей датой и решает
 *   доступна часть или показывать «Откроется DD числа»
 *
 * Закреплённое сообщение (только Анна, 1 шт на клуб — см. AI-CONTEXT v5):
 * - pinnedMessageId — ссылка на ChatMessage с isPinned=true
 *
 * Параллельные клубы:
 * - Активный клуб определяется по isActive=true (выставляется cron-задачей
 *   ежедневно в 00:00) или фильтром startsAt <= now < endsAt
 * - В момент перехода с месяца на месяц предыдущий имеет archiveUntilDate
 *   в будущем — это используется в middleware requireSubscription для решения
 *   допускать ли истёкшую подписку к старому клубу
 */
const partScheduleSchema = new mongoose.Schema(
  {
    partNumber: { type: Number, required: true },
    opensAt: { type: Date, required: true },
  },
  { _id: false }
);

const clubMonthSchema = new mongoose.Schema(
  {
    // Идентификация месяца
    month: { type: Number, required: true, min: 1, max: 12 },
    year: { type: Number, required: true },

    // Книга разбора
    bookId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'Book',
      required: true,
    },

    // Дублирование данных из Book для скорости (denormalization).
    // Заполняется при создании клуба, обновляется если книга меняется.
    title: { type: String, required: true }, // 'Бегство от свободы'
    author: { type: String, default: '' },

    // Расписание клуба
    startsAt: { type: Date, required: true },
    endsAt: { type: Date, required: true },
    archiveUntilDate: { type: Date, required: true }, // endsAt + 21 день

    // Расписание открытия частей (если пусто — все части доступны со старта)
    partSchedule: { type: [partScheduleSchema], default: [] },

    // Состояние
    isActive: { type: Boolean, default: false, index: true },

    // Статистика (обновляется триггерами в сервисах)
    participantCount: { type: Number, default: 0 },
    messageCount: { type: Number, default: 0 },

    // Закреплённое сообщение (1 шт, может закрепить только Анна)
    pinnedMessageId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'ChatMessage',
      default: null,
    },
  },
  {
    timestamps: true,
  }
);

// Уникальная пара (month, year) — один клуб на один месяц.
clubMonthSchema.index({ year: 1, month: 1 }, { unique: true });

// Быстрый поиск активного клуба.
clubMonthSchema.index({ isActive: 1, startsAt: -1 });

// Для определения архивного доступа в middleware (4.2).
clubMonthSchema.index({ endsAt: 1, archiveUntilDate: 1 });

module.exports = mongoose.model('ClubMonth', clubMonthSchema);
